README.md
Language: [English](README.md) | [中文](README_zh-CN.md)

# simple_update
![Pub Version](https://img.shields.io/pub/v/simple_update?style=flat-square)
![Platform](https://img.shields.io/badge/platform-flutter%7Cflutter%20web%7Cdart%20vm-brightgreen)

The simplest way to update app, support all platforms.

By default, the new version would be downloaded from [avenge.app](https://avenge.app), which is a simple Version-Management-System, welcome to try ^_^

You can set your own server to publish apps.

## Getting Started

1. Register for free

   [https://avenge.app/register](https://avenge.app/register)
2. Create app & create version

   [https://avenge.app/developer/resources/app-apps/new](https://avenge.app/developer/resources/app-apps/new)
   
   [https://avenge.app/developer/resources/app-versions/new](https://avenge.app/developer/resources/app-versions/new)

3. Install
   ```yaml
   dependencies:
       simple_update: ^2.0.12
   ```
   
   
4. Usage

   This is an example of Android. 
   
   In this example, you can use **[ota_update](https://pub.dev/packages/ota_update)** to update your app.
   
   ```dart
   import 'package:ota_update/ota_update.dart';
   import 'package:simple_update/simple_update.dart' as simple;
   import 'package:package_info/package_info.dart';
   
   code...
   RaisedButton(
      onPressed: () async {
        var updater = new simple.SimpleUpdate(
            appId: 1,
            appKey: 'g4rehwe8fq4qe9rgh4q123');
        var info = await PackageInfo.fromPlatform();
        var res = await updater.checkUpdate(
            platform: simple.Platform.Android,
            info: simple.Info(
                appName: info.appName,
                version: info.version,
                buildNumber: int.parse(info.buildNumber)));
        if (res == true) {
          var re = await showDialog<bool>(
            context: context,
            builder: (context) => AlertDialog(
              title: Text('New version'),
              content: Text('Install the new version?'),
              actions: [
                FlatButton(
                    onPressed: () => Navigator.pop(context),
                    child: Text('Cancel')),
                FlatButton(
                    onPressed: () => Navigator.pop(context, true),
                    child: Text('Yes')),
              ],
            ),
          );
          if (re == true) {
            var r = await updater.update(
                platform: simple.Platform.Android,
                downloader: (info, version) {
                  return OtaUpdate()
                      .execute(
                    version.src,
                    destinationFilename:
                        '${info.appName}_${version.name}+${version.number}.apk',
                    sha256checksum: version.sha256,
                  )
                      .transform(StreamTransformer<OtaEvent,
                              simple.Event>.fromHandlers(
                          handleData: (event, sink) {
                    sink.add(simple.Event(
                        status: simple.Status.values[event.status.index],
                        value: event.value));
                  }));
                });
            if (r is Stream<simple.Event>) {
              r.listen((event) {
                switch (event.status) {
                  case simple.Status.DOWNLOADING:
                    print('downloading %${event.value}');
                    break;
                  case simple.Status.INSTALLING:
                    print('installing');
                    break;
                  case simple.Status.ALREADY_RUNNING_ERROR:
                    print('download is already running');
                    break;
                  case simple.Status.PERMISSION_NOT_GRANTED_ERROR:
                    print(
                        'could not continue because of missing permissions');
                    break;
                  case simple.Status.INTERNAL_ERROR:
                  case simple.Status.DOWNLOAD_ERROR:
                  case simple.Status.CHECKSUM_ERROR:
                    print('error ${event.value}');
                }
              });
            }
          }
        } else {
          print('no latest version');
        }
      },
      child: Text('Update'),
   )
   code...
   ```

## Build publish server

  
   The default value of [SimpleUpdate.apiPrefix] is 'https://avenge.app/api', you can set your own server to publish apps.
   ```dart
   var updater = SimpleUpdate(apiPrefix: 'your own server');
   ```
  
  When request the latest version, it should send these parameters to remote server by GET method:
  > https://avenge.app/api/latest?app_id=456&app_key=abc123&platform=0

  platform: 0:android 1:ios 2:fuchsia 3:linux 4:windows 5:macOS

  Then, the remote server will send back a json:
  ```json
  {
    "code": 0,
    "message": "success",
    "data": {
     "id": 123,
     "app_id": 456,
     "platform": 0,
     "name": "1.0.1",
     "number": 24,
     "src": "file url",
     "sha256": "sha256 of file",
     "created_at": "2020-07-29 12:15:46"
    }
  }
  ```
   If *code* is 0, it means success.
  
   If something went wrong, it would be:
   ```json
   {
     "code": -1,
     "message": "something wrong ..."
   }
   ```

## All projects
| Plugins                                                      | Status                                                       | Description                                                  |
| ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| [simple_log](https://github.com/creatint/flutter_simple_log) | ![Pub Version](https://img.shields.io/pub/v/simple_log?style=flat-square) | The simplest way to upload logs to remote server, support all platforms |
| [simple_update](https://github.com/creatint/flutter_simple_update) | ![Pub Version](https://img.shields.io/pub/v/simple_update?style=flat-square) | The simplest way to update your app, support all platforms |


CHANGELOG.md
## 2.0.12

* Change README.md and README_zh-CN.md

## 2.0.11

* Change README.md and README_zh-CN.md

## 2.0.10

* Change README.md

## 2.0.9

* Change the default domain of api.

## 2.0.8

* Format code

## 2.0.7

* Modify addDownloadCount()

## 2.0.6

* Add addDownloadCount()

## 2.0.5

* Modify the Info's API

## 2.0.4

* Modify the PackageInfo's API

## 2.0.3

* Fix README.md

## 2.0.2

* Fix README.md

## 2.0.1

* Fix README.md

## 2.0.0

* Change to dart package, support all platforms.

## 1.0.0

* First version, only support Android


test/simple_update_test.dart
import 'package:simple_update/simple_update.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:simple_update_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is Text && widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


example/lib/home.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:ota_update/ota_update.dart';
import 'package:simple_update/simple_update.dart' as simple;
import 'package:package_info/package_info.dart';

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  simple.SimpleUpdate updater;
  bool downloading = false;

  @override
  void initState() {
    super.initState();
    updater = simple.SimpleUpdate(
        appId: 1,
        appKey: 'g4rehwe8fq4qe9rgh4q',
        apiPrefix: 'http://avenge.app/api');
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('SimpleUpdate'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            RaisedButton(
              onPressed: () async {
                if (downloading) {
                  return;
                }
                var info = await PackageInfo.fromPlatform();
                var res = await updater.checkUpdate(
                    platform: simple.Platform.Android,
                    info: simple.Info(
                        appName: info.appName,
                        version: info.version,
                        buildNumber: int.parse(info.buildNumber)));
                if (res == true) {
                  var re = await showDialog<bool>(
                    context: context,
                    builder: (context) => AlertDialog(
                      title: Text('New version'),
                      content: Text('Install the new version?'),
                      actions: [
                        FlatButton(
                            onPressed: () => Navigator.pop(context),
                            child: Text('Cancel')),
                        FlatButton(
                            onPressed: () => Navigator.pop(context, true),
                            child: Text('Yes')),
                      ],
                    ),
                  );
                  if (re == true) {
                    var r = await updater.update(
                        platform: simple.Platform.Android,
                        downloader: (info, version) {
                          print(version.src);
                          print(
                              'download123 file ${info.appName}_${version.name}+${version.number}.apk');
                          return OtaUpdate()
                              .execute(
                            version.src,
                            destinationFilename:
                                '${info.appName}_${version.name}+${version.number}.apk',
                            sha256checksum: version.sha256,
                          )
                              .transform(StreamTransformer<OtaEvent,
                                      simple.Event>.fromHandlers(
                                  handleData: (event, sink) {
                            sink.add(simple.Event(
                                status:
                                    simple.Status.values[event.status.index],
                                value: event.value));
                          }));
                        });
                    if (r is Stream<simple.Event>) {
                      r.listen((event) {
                        switch (event.status) {
                          case simple.Status.DOWNLOADING:
                            print('downloading %${event.value}');
                            downloading = true;
                            break;
                          case simple.Status.INSTALLING:
                            print('installing');
                            downloading = false;
                            break;
                          case simple.Status.ALREADY_RUNNING_ERROR:
                            print('download is already running');
                            downloading = false;
                            break;
                          case simple.Status.PERMISSION_NOT_GRANTED_ERROR:
                            print(
                                'could not continue because of missing permissions');
                            downloading = false;
                            break;
                          case simple.Status.INTERNAL_ERROR:
                          case simple.Status.DOWNLOAD_ERROR:
                          case simple.Status.CHECKSUM_ERROR:
                            print('error ${event.value}');
                            downloading = false;
                        }
                      });
                    }
                  }
                } else {
                  print('no latest version');
                }
              },
              child: Text('Update'),
            )
          ],
        ),
      ),
    );
  }
}


lib/src/simple_update_base.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;

enum Platform { Android, IOS, Fuchsia, Linux, Windows, MacOS }

typedef Downloader = Stream<Event> Function(Info info, Version version);

class SimpleUpdate {
  String _apiPrefix;
  int _appId;
  String _appKey;
  http.Client _client;
  Version _version;
  Info _info;

  static final Map<String, SimpleUpdate> _cache = {};

  /// Get singleton instance by [key]
  ///
  /// ```dart
  /// var updater = SimpleUpdater();
  /// ```
  ///
  /// The default value of [apiPrefix] is *https://avenge.app/api*, you can set your own server to publish apps.
  /// ```dart
  /// var updater = SimpleUpdate(apiPrefix: 'your own server');
  /// ```
  ///
  /// When request the latest version, it should send these parameters to remote server:
  /// >> https://avenge.app/api/latest?app_id=[_appId]&app_key=[_appKey]&platform=0
  ///
  /// platform 0:android 1:ios 2:fuchsia 3:linux 4:windows 5:macOS
  ///
  ///
  /// Then, the remote server will send back a json:
  /// ```json
  /// {
  ///   "code": 0,
  ///   "message": "success",
  ///   "data": {
  ///    "id": 123,
  ///    "app_id": 456,
  ///    "platform": 0,
  ///    "name": "1.0.1",
  ///    "number": 24,
  ///    "src": "https://avenge.app/storage/apps/2020-07/app_name_1.0.1+24.apk",
  ///    "sha256": "sha256 of file",
  ///    "created_at": "2020-07-29 12:15:46"
  ///   }
  /// }
  /// ```
  /// If *code* is 0, it means success.
  ///
  /// If something went wrong, it would be:
  /// ```json
  /// {
  ///   "code": -1,
  ///   "message": "something wrong ..."
  /// }
  /// ```
  factory SimpleUpdate(
      {String key = 'default',
      int appId,
      String appKey,
      String user,
      String flag,
      String apiPrefix = 'https://avenge.app/api'}) {
    if (_cache.containsKey(key)) {
      _cache[key]
        .._appId = appId
        .._appKey = appKey
        .._apiPrefix = apiPrefix;
      return _cache[key];
    }
    _cache[key] = SimpleUpdate._internal(
        appId: appId,
        appKey: appKey,
        user: user,
        flag: flag,
        apiPrefix: apiPrefix);
    return _cache[key];
  }

  SimpleUpdate._internal(
      {int appId, String appKey, String user, String flag, String apiPrefix}) {
    _appId = appId;
    _appKey = appKey;
    _apiPrefix = apiPrefix;
    _client = http.Client();
  }

  /// Get the latest version
  ///
  /// If return null, it means there's no latest version
  Future<Version> getLatest(Platform platform) async {
    var res =
        await _client.get('$_apiPrefix/latest?app_id=$_appId&app_key=$_appKey');

    if (res.statusCode != 200) {
      return null;
    }

    if (res.body == null || res.body.isEmpty) {
      return null;
    }

    var data = jsonDecode(res.body);

    if (data == null) {
      return null;
    }

    return Version.fromJson(data['data']);
  }

  /// Check if the app has new version
  Future<bool> checkUpdate({Platform platform, Info info}) async {
    var version = await getLatest(platform);

    if (version != null && version.number > info.buildNumber) {
      _version = version;
      _info = info;
      return true;
    }

    return false;
  }

  /// Update to the latest version
  ///
  /// only support Android
  /// return String or Stream<Event>
  Future<dynamic> update({Platform platform, Downloader downloader}) async {
    if (platform == Platform.Android) {
      if (_version == null) {
        var version = await getLatest(platform);
        if (version == null) {
          return 'No latest version';
        }
        _version = version;
      }

      if (_version.src == null || _version.src.isEmpty) {
        return 'No resource';
      }

      try {
        addDownloadCount();
        return downloader(_info, _version);
      } catch (e) {
        return 'Failed to make update. Details: $e';
      }
    }
  }

  void addDownloadCount() {
    _client.get(
        '$_apiPrefix/addDownloadCount?app_id=$_appId&app_key=$_appKey&version_id=${_version.id}');
  }
}

/// Version of APP
class Version {
  final int id;
  final int appId;
  final Platform platform;
  final String name;
  final int number;
  final String src;
  final String sha256;
  final DateTime createdAt;

  Version(this.id, this.appId, this.platform, this.name, this.number, this.src,
      this.sha256, this.createdAt);

  Version.fromJson(json)
      : assert(json['created_at'] != null),
        id = json['id'],
        appId = json['app_id'],
        platform = Platform.values[json['platform']],
        name = json['name'],
        number = json['number'],
        src = json['src'],
        sha256 = json['sha256'],
        createdAt = DateTime.parse(json['created_at']);

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'app_id': appId,
      'platform': platform,
      'name': name,
      'number': number,
      'src': src,
      'sha256': sha256,
      'created_at': createdAt.toString().substring(0, 19)
    };
  }

  @override
  String toString() {
    return toJson().toString();
  }
}

///Event describing current status
class Event {
  Event({this.status, this.value});

  /// Current status as enum value
  Status status;

  /// Additional status info e.g. percents downloaded or error message (can be null)
  String value;
}

/// Enum values describing states
enum Status {
  /// FILE IS BEING DOWNLOADED
  DOWNLOADING,

  /// INSTALLATION HAS BEEN TRIGGERED
  INSTALLING,

  /// DOWNLOAD IS ALREADY RUNNING
  ALREADY_RUNNING_ERROR,

  /// COULD NOT CONTINUE BECAUSE OF MISSING PERMISSIONS
  PERMISSION_NOT_GRANTED_ERROR,

  /// UNKWNON ERROR. SEE VALUE FOR MORE INFROMATION
  INTERNAL_ERROR,

  /// FILE COULD NOT BE DOWNLOADED. SEE VALUE FOR MORE INFORMATION
  DOWNLOAD_ERROR,

  /// CHECKSUM VERIFICATION FAILED. MOSTLY THIS IS DUE INCORRECT OR CORRUPTED FILE
  /// THIS IS ALSO RETURNED IF PLUGIN WAS UNABLE TO CALCULATE SHA 256 HASH OF DOWNLOADED FILE
  /// SEE VALUE FOR MORE INFORMATION
  CHECKSUM_ERROR
}

class Info {
  Info({
    this.appName,
    this.version,
    this.buildNumber,
  });

  final String appName;

  final String version;

  final int buildNumber;
}
